fpga_interchange/populate_chip_info.py

```diff
@@ -1740,6 +1759,438 @@ def get_cell(cell_type):
+                        graph[sink].append((source, 0))
+
+            cluster = {}
+            cluster['name'] = macro_name
+
+            if debug:
+                print("\tChecking connections graph")
+
+            # Check if connection graph is a tree
+            subgraphs, cycle = check(graph)
+            assert not cycle, "Code is not design to cope with cycles inside macros"
+            root = None
+            sources = None
+            if subgraphs == 1:
+                # If grah is a tree, search for node, such that when it's remved ther is no connection from any source to any sink
+                root = search_for_root(graph)
+                if root is None:
+                    root = source_cells(graph)[0]
+                cluster['root_cell_types'] = [cell_instances[root].cell_name]
+            else:
+                root = source_cells(graph)[0]
+                cluster['root_cell_types'] = [cell_instances[root].cell_name]
+
+            cluster['connection_graph'], cell_idx_map = find_connection_graph(
+                old_nets, cell_instances, cell_pin_direction_map,
+                [VCC_cell, GND_cell], root)
+            if cluster['connection_graph'] is None:
+                continue
+
+            cluster['out_of_site_clusters'] = check_if_out_site(
+                cell_instances, cell_site_type_map, device,
+                cell_site_type_to_bels_map, cluster['connection_graph'], root)
+
+            cluster['phy_graph'] = None
+            if macro_name in phy_placements:
+                cluster['phy_graph'] = []
+                for site, places in phy_placements[macro_name].items():
+                    site_placements = []
+                    length = len(places[list(cell_instances.keys())[0]])
+                    for i in range(length):
+                        l = [None] * len(cell_instances)
+                        for cell in cell_instances.keys():
+                            l[cell_idx_map[cell]] = places[cell][i]
+                        l = tuple(l)
+                        site_placements.append(l)
+                    cluster['phy_graph'].append(tuple([site, site_placements]))
+
+            cluster["disallow_other_cells"] = False
+            cluster['chainable_ports'] = []
+            cluster['from_macro'] = True
+            cluster['required_cells'] = []
+            count_cell_type_instances = {}
+
+            if debug:
+                print("\tListing required cells")
+
+            for cell_name, cell_data in cell_instances.items():
+                if cell_data.cell_name not in count_cell_type_instances:
+                    count_cell_type_instances[cell_data.cell_name] = 0
+                count_cell_type_instances[cell_data.cell_name] += 1
+            for cell_type, count in count_cell_type_instances.items():
+                cluster['required_cells'].append((cell_type, count))
+            if cluster['phy_graph'] is not None:
+                clusters.append(cluster)
```

Comment:
We might move the check on whether the macro is present in the `phy_placement` already in the beginning and avoid arriving at the end of the loop to realize that the macro should not have been expanded into a cluster.